### PR TITLE
RP metrics - frontend 

### PR DESCRIPTION
**Pull request by mjudeikis**
Add statsd metrics for RP frontend

### CODE REVIEW

**Comment by mjudeikis** [comment]
FrontEnd metrics example:
```
"Metric":"frontend.duration.GET.getOpenShiftClusters.2019-12-31-preview.200.unknown","Account":"metrics-account","Namespace":"metrics-namespace","Dims":{"location":"eastus"},"TS":"2020-01-06T17:42:54.796"}:81.000000|f

{"Metric":"frontend.duration.GET.getOpenShiftCluster.2019-12-31-preview.404.unknown","Account":"metrics-account","Namespace":"metrics-namespace","Dims":{"location":"eastus"},"TS":"2020-01-06T17:43:25.720"}:82.000000|f
```

Not sure If I like this overloaded metric format. 
@jim-minter wdyt? Last commits adds them

File: pkg/frontend/frontend.go [hunk 1]
```diff
@@ -47,20 +49,23 @@ type Runnable interface {
...
 		db:      db,
+		m:       m,
 	}
 
 	l, err := f.env.Listen()
 	if err != nil {
 		return nil, err
 	}
 
+	m.EmitFloat("dummy", float64(5))
```
Comment (mjudeikis) [inline pkg/frontend/frontend.go]: Will remove

File: env.example [hunk 2]
```diff
@@ -1,3 +1,8 @@
 export LOCATION=eastus
 
+export METRICS_SOCKET=/tmp/aro-metrics.socket
```
Comment (jim-minter) [inline env.example]: can we get down to 0 configuration for dev?
e.g. hard-code the metrics socket to mdm_statsd.socket (in prod, we'll have Docker mount it in place)
and avoid requirement for the others to be set.

File: hack/metrics/server.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
+package main
```
Comment (jim-minter) [inline hack/metrics/server.go]: can you use `nc -U -l mdm_statsd.socket` and avoid the need for this file?

File: pkg/env/dev.go [hunk 4]
```diff
@@ -284,3 +284,11 @@ func (d *dev) CreateARMResourceGroupRoleAssignment(ctx context.Context, fpAuthor
 		return true, nil
 	})
 }
+
+func (d *dev) MetricsSocket() string {
```
Comment (jim-minter) [inline pkg/env/dev.go]: I think both of these are avoidable.

File: pkg/frontend/frontend.go [hunk 5]
```diff
@@ -228,7 +234,8 @@ func (f *frontend) Run(stop <-chan struct{}, done chan<- struct{}) {
 	}
 }
 
-func reply(log *logrus.Entry, w http.ResponseWriter, header http.Header, b []byte, err error) {
+func (f *frontend) reply(ctx context.Context, log *logrus.Entry, w http.ResponseWriter, header http.Header, b []byte, err error) {
+	defer f.emitFrontendMetric(ctx, err)
```
Comment (jim-minter) [inline pkg/frontend/frontend.go]: can this go into the middleware?

File: pkg/frontend/metrics.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/frontend/middleware"
+	"github.com/davecgh/go-spew/spew"
+)
+
+func (f *frontend) emitFrontendMetric(ctx context.Context, err error) {
```
Comment (jim-minter) [inline pkg/frontend/metrics.go]: can this go into the middleware? (after the call to `h.ServeHTTP(w, r)`)

File: pkg/frontend/middleware/middleware.go [hunk 7]
```diff
@@ -9,4 +9,5 @@ const (
 	ContextKeyLog contextKey = iota
 	ContextKeyOriginalPath
 	ContextKeyBody
+	ContextKeyMetrics
```
Comment (jim-minter) [inline pkg/frontend/middleware/middleware.go]: why is this needed?

File: pkg/metrics/metrics.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
+package metrics
+
+type Option func(*ClientConfig)
```
Comment (jim-minter) [inline pkg/metrics/metrics.go]: this Option mechanism seems overly heavyweight

File: pkg/deploy/rp.go [hunk 9]
```diff
@@ -189,7 +189,8 @@ func (g *generator) vmss() *arm.Resource {
 		fmt.Sprintf("base64ToString('%s')", base64.StdEncoding.EncodeToString([]byte("set -ex\n\n"))),
 	}
 
-	for _, variable := range []string{"pullSecret", "rpImage", "rpImageAuth"} {
+	for _, variable := range []string{"pullSecret", "rpImage", "rpImageAuth",
```
Comment (jim-minter) [inline pkg/deploy/rp.go]: probably worth doing one per line, sorted

File: pkg/metrics/statsd/metrics.go [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+	Namespace string
+	Dims      map[string]string
+	TS        time.Time
+}
+
+// Float point number is used to represent floating point numbers which vary a
+// lot in its magnitude and thus warrant storage in a mantissa + expontent format
+// (such as physical quantities, results of complex calculations, etc). <value>
+// can be any fractional number, which would be stored as double precision IEEE 754 float.
+type float struct {
```
Comment (jim-minter) [inline pkg/metrics/statsd/metrics.go]: there's a lot of duplication in this file to say that the only meaningful difference I can see is `:%d|g` or `:%f|f`

File: pkg/metrics/statsd/statsd.go [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+	}
+	_, err = c.Send(b)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Send data to statsd daemon
+func (c *clientConfig) Send(data []byte) (int, error) {
```
Comment (jim-minter) [inline pkg/metrics/statsd/statsd.go]: calling this Write would be more customary

File: pkg/metrics/metrics.go [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+// Interface represents metrics interface
+type Interface interface {
+	Clone(opts ...Option) Interface
+	Send([]byte) (int, error)
+	EmitFloat(string, float64) error
+	EmitFloatWithDims(string, float64, ...string) error
+	EmitGauge(string, int64) error
+	EmitGaugeWithDims(string, int64, ...string) error
+}
```
Comment (jim-minter) [inline pkg/metrics/metrics.go]: I think this interface can be pared right back to `Emit{Float,Gauge}(metric string, value whatever, dims map[string]string) error - then you sub in namespace, accounts, ts, marshal lock and write.

I don't think that Option, Prefix or Dimensions are needed.

Without the prefix logic, code that just says Emit("frontend.duration") it is simpler, and it is more greppable.

File: pkg/metrics/statsd/statsd.go [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+		config: &config{
+			socket: env.MetricsSocket(),
+			env:    env,
+		},
+		ClientConfig: *cConfig,
+	}
+
+	var err error
+	clientConfig.config.conn, err = net.Dial("unix", clientConfig.config.socket)
+	if err != nil && env.IsDevelopment() {
```
Comment (jim-minter) [inline pkg/metrics/statsd/statsd.go]: `_, ok := _env.(env.Dev); ok`
Specifically only ignore ENOENT in development mode.

File: pkg/frontend/metrics.go (hunk 6 above)
Comment (mjudeikis) [reply pkg/frontend/metrics.go]: How would you get the Internal Error code or Response code? If we would track the only Duration, this would work.

File: pkg/frontend/frontend.go (hunk 5 above)
Comment (mjudeikis) [reply pkg/frontend/frontend.go]: How? Error codes are response codes?

File: pkg/metrics/statsd/statsd.go (hunk 11 above)
Comment (mjudeikis) [reply pkg/metrics/statsd/statsd.go]: explain, please?

File: pkg/frontend/middleware/metric.go [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/gorilla/mux"
+
+	"github.com/Azure/ARO-RP/pkg/metrics"
+)
+
+type frontendMetric struct {
```
Comment (jim-minter) [inline pkg/frontend/middleware/metric.go]: not used?

File: pkg/metrics/statsd/statsd.go [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+// New method to initialize udp connection
+func New(ctx context.Context, log *logrus.Entry, _env env.Interface) (metrics.Interface, error) {
+	// defaults- dev
+	config := &Statsd{
+		account:   os.Getenv("METRICS_ACCOUNT"),
+		namespace: os.Getenv("METRICS_NAMESPACE"),
+		dims: map[string]string{
+			"location": _env.Location(),
```
Comment (jim-minter) [inline pkg/metrics/statsd/statsd.go]: if we have the mdsd config add this, you won't need AppendDimensions and can straight use a dimensions map passed in by the caller.

File: pkg/metrics/statsd/statsd.go [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+	namespace string
+	dims      map[string]string
+
+	socket string
+	conn   net.Conn
+	env    env.Interface
+	mu     sync.Mutex
+
+	// for testing
+	Muted bool
```
Comment (jim-minter) [inline pkg/metrics/statsd/statsd.go]: a pkg/metrics/noop which implements the metrics interface but does nothing is preferable.

File: pkg/metrics/statsd/metrics.go [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+		Metric:    f.Metric,
+		Account:   f.Account,
+		Namespace: f.Namespace,
+		Dims:      f.Dims,
+		TS:        f.TS.UTC().Format("2006-01-02T15:04:05.000"),
+	})
+}
+
+// Marshal a Float into its statsd format.  Call this instead of MarshalJSON().
+func (f *metric) Marshal() ([]byte, error) {
```
Comment (jim-minter) [inline pkg/metrics/statsd/metrics.go]: could consider calling this MarshalStatsd

File: pkg/metrics/statsd/statsd.go [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+	return config, nil
+}
+
+// Close closes the connection
+func (c *Statsd) Close() {
+	c.conn.Close()
+}
+
+// EmitFloat records float information
+func (c *Statsd) EmitFloat(stat string, value float64, dims ...string) error {
```
Comment (jim-minter) [inline pkg/metrics/statsd/statsd.go]: can you further reduce the duplication between EmitFloat and EmitGauge?

File: pkg/metrics/statsd/statsd.go [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+	}
+	_, err = c.Send(b)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Send data to statsd daemon
+func (c *Statsd) Send(data []byte) (int, error) {
```
Comment (jim-minter) [inline pkg/metrics/statsd/statsd.go]: this should reduce to:
```
func (c *Statsd) Write(b []byte) (int, error) {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.conn.Write(b)
}
```

File: pkg/metrics/statsd/statsd.go [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/metrics"
+)
+
+// statsd implementation for https://genevamondocs.azurewebsites.net/collect/references/statsdref.html
+
+// Now stubs time.Now
+var Now = time.Now
```
Comment (jim-minter) [inline pkg/metrics/statsd/statsd.go]: better inside the Statsd struct

File: pkg/metrics/metrics.go [hunk 21]
```diff
@@ -0,0 +1,11 @@
+package metrics
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+// Interface represents metrics interface
+type Interface interface {
+	Close()
+	EmitFloat(string, float64, ...string) error
```
Comment (jim-minter) [inline pkg/metrics/metrics.go]: instead of ...string, why not map[string]string?

File: pkg/metrics/statsd/statsd.go [hunk 22]
```diff
@@ -0,0 +1,152 @@
...
+		account:   os.Getenv("METRICS_ACCOUNT"),
+		namespace: os.Getenv("METRICS_NAMESPACE"),
+		dims: map[string]string{
+			"location": _env.Location(),
+		},
+		socket: defaultSocket,
+	}
+
+	var err error
+	config.conn, err = net.Dial("unix", config.socket)
```
Comment (jim-minter) [inline pkg/metrics/statsd/statsd.go]: why not hard code to "mdm_statsd.socket" (i.e. in the current working directory?) (and add to .gitignore)

File: pkg/frontend/middleware/metric.go (hunk 14 above)
Comment (mjudeikis) [reply pkg/frontend/middleware/metric.go]: Deleted, not yet commited

File: pkg/metrics/statsd/statsd.go (hunk 15 above)
Comment (mjudeikis) [reply pkg/metrics/statsd/statsd.go]: yes. I would like to tackle this as later on when I start integrating with mdsd hands on.

File: pkg/metrics/metrics.go (hunk 21 above)
Comment (mjudeikis) [reply pkg/metrics/metrics.go]: It looks and feelsmuch nicer and cleaner when constructing metrics calls:
```
defer m.EmitFloat("frontend.duration",
				float64(time.Now().Sub(t)/time.Millisecond),
				"verb", r.Method,
				"api-version", vars["api-version"],
				"code", strconv.Itoa(w.(*logResponseWriter).statusCode),
				"route", routeName)
```
And works just the same. 
```
defer m.EmitFloat("frontend.duration",float64(time.Now().Sub(t)/time.Millisecond))
```

File: pkg/deploy/rp.go [hunk 23]
```diff
@@ -237,6 +243,8 @@ fi
 cat >/etc/sysconfig/arorp <<EOF
 RP_IMAGE='$RPIMAGE'
 PULL_SECRET='$PULLSECRET'
+METRICS_NAMESPACE='$METRICSNAMESPACE'
+METRICS_ACCOUNT='$METRICSACCOUNT'
```
Comment (jim-minter) [inline pkg/deploy/rp.go]: will need to add `-e METRICS_ACCOUNT -e METRICS_NAMESPACE` on line 259

File: pkg/metrics/statsd/statsd.go [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+// EmitGauge records gauge information
+func (c *Statsd) EmitGauge(stat string, value int64, dims map[string]string) error {
+	return c.emitMetric(metric{
+		Metric:     stat,
+		Dims:       dims,
+		ValueGauge: to.Int64Ptr(int64(value)),
+	})
+}
+
+func (c *Statsd) emitMetric(m metric) error {
```
Comment (jim-minter) [inline pkg/metrics/statsd/statsd.go]: `m *metric`

File: pkg/metrics/statsd/statsd.go [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+	defaultSocket = "mdm_statsd.socket"
+)
+
+// Statsd defines internal statsd client
+// It should be cloned, but not modified
+type Statsd struct {
+	account   string
+	namespace string
+
+	conn net.Conn
```
Comment (jim-minter) [inline pkg/metrics/statsd/statsd.go]: writing `conn io.WriteCloser` here will simplify your test

File: pkg/metrics/statsd/statsd_test.go [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+	return c.err
+}
+
+func newTestClient() (*Statsd, error) {
+	return &Statsd{
+		conn:      &testConn{},
+		account:   "test_account",
+		namespace: "test_namespace",
+		now: func() time.Time {
+			time, _ := time.Parse("", "0001-01-01T00:00:00.000")
```
Comment (jim-minter) [inline pkg/metrics/statsd/statsd_test.go]: nit: `return time.Time{}` is also feasible

File: pkg/frontend/middleware/metric.go [hunk 27]
```diff
@@ -0,0 +1,53 @@
...
+			} else {
+				routeName = "unknown"
+			}
+
+			w = &logResponseWriter{ResponseWriter: w, statusCode: http.StatusOK}
+
+			h.ServeHTTP(w, r)
+
+			// request count
+			defer m.EmitGauge("frontend.count",
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply pkg/frontend/middleware/metric.go]: I tested this case:
1. Count is before serve
2. delete cluster panics
3. I get 200 on Count and 500 on the duration

Both after serve returns 500.

File: pkg/metrics/statsd/statsd.go [hunk 28]
```diff
@@ -0,0 +1,106 @@
...
+	mu   sync.Mutex
+
+	now func() time.Time
+}
+
+// New method to initialize udp connection
+func New(ctx context.Context, log *logrus.Entry, _env env.Interface) (metrics.Interface, error) {
+	// defaults- dev
+	config := &Statsd{
+		account:   os.Getenv("METRICS_ACCOUNT"),
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply pkg/metrics/statsd/statsd.go]: good catch. forgot about that part of the code.

File: pkg/deploy/rp.go [hunk 29]
```diff
@@ -248,7 +256,9 @@ Requires=docker.service
 EnvironmentFile=/etc/sysconfig/arorp
 ExecStartPre=-/usr/bin/docker rm -f %n
 ExecStartPre=/usr/bin/docker pull \$RP_IMAGE
-ExecStart=/usr/bin/docker run --rm --name %n -p 443:8443 -e PULL_SECRET \$RP_IMAGE rp
+ExecStart=/usr/bin/docker run --rm --name %n -p 443:8443 -e PULL_SECRET \$RP_IMAGE \
+							  -e METRICS_ACCOUNT \$METRICS_ACCOUNT \
+							  -e METRICS_NAMESPACE \$METRICS_NAMESPACE  rp
```
Comment (jim-minter) [inline pkg/deploy/rp.go]: `ExecStart=/usr/bin/docker run --rm --name %n -p 443:8443 -e METRICS_ACCOUNT -e METRICS_NAMESPACE -e PULL_SECRET \$RP_IMAGE rp`

File: pkg/frontend/middleware/metrics.go [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+			} else {
+				routeName = "unknown"
+			}
+
+			w = &logResponseWriter{ResponseWriter: w, statusCode: http.StatusOK}
+
+			h.ServeHTTP(w, r)
+
+			// request count
+			defer m.EmitGauge("frontend.count",
```
Comment (jim-minter) [inline pkg/frontend/middleware/metrics.go]: defer must be above ServeHTTP

File: pkg/metrics/statsd/statsd.go [hunk 31]
```diff
@@ -0,0 +1,109 @@
...
+
+const defaultSocket = "mdm_statsd.socket"
+
+// Statsd defines internal statsd client
+// It should be cloned, but not modified
+type Statsd struct {
+	account   string
+	namespace string
+
+	conn io.ReadWriteCloser
```
Comment (jim-minter) [inline pkg/metrics/statsd/statsd.go]: io.WriteCloser
